rendermodules/rough_align/apply_rough_alignment_to_montages.py

```diff
@@ -61,39 +66,157 @@
+
+    results = {}
+
+    if read_masks_from_lowres_stack:
+        for t in tilespecs:
+            if t.ip[0].maskUrl is not None:
+                results[t.tileId] = t.ip[0].maskUrl
+
+    elif mask_input_dir is not None:
+        tids = [t.tileId for t in tilespecs]
+        maskfiles = []
+        for ext in exts:
+            e = mask_input_dir + '/*.' + ext
+            maskfiles += glob.glob(e)
+        maskbasenames = np.array([os.path.basename(os.path.splitext(p)[0])
+                                  for p in maskfiles])
+        for t in tids:
+            ind = np.argwhere(maskbasenames == t).flatten()
+            if ind.size > 1:
+                raise RenderModuleException("more than one mask in "
+                                            "directory %s matches tileId %s" %
+                                            (mask_input_dir, t))
+            if ind.size == 0:
+                continue
+            results[t] = pathlib.Path(maskfiles[ind[0]]).as_uri()
+
+    return results
+
+
+def add_masks_to_lowres(render, stack, z, mask_map):
+    if len(mask_map) == 0:
+        return
+
+    tilespecs = renderapi.tilespec.get_tile_specs_from_z(
+            stack, z, render=render)
+
+    for t in tilespecs:
+        if t.tileId in mask_map:
+            t.ip['0'].maskUrl = mask_map[t.tileId]
+            renderapi.client.import_tilespecs(
+                    stack,
+                    [t],
+                    render=render)
+    return
+
+
+def filter_highres_with_masks(resolved_highres, tspec_lowres, mask_map):
+    """function to return a filtered list of tilespecs from a
+       ResolvedTiles object, based on a lowres mask
+
+    Parameters
+    ----------
+    resolved_highres : renderapi.resolvedtiles.ResolvedTiles object
+        tilespecs and transforms from a single section
+    tspec_lowres: renderapi.tilespec.TileSpec object
+        tilespec from a downsampled stack
+    mask_map: dict
+        keys should match lowres tileids, values are mask file URI
+
+    Returns
+    -------
+    new_highres: List of renderapi.tilespec.TileSpec objects
+        which highres specs are fully contained
+        within the boundary of mask=255
+    """
+
+    if tspec_lowres.tileId not in mask_map.keys():
+        return resolved_highres.tilespecs
+
+    impath = urllib.parse.unquote(
+                 urllib.parse.urlparse(
+                     mask_map[tspec_lowres.tileId]).path)
+    maskim = cv2.imread(impath, 0)
+    maskim = 255 - maskim
+    _, contours, _ = cv2.findContours(
+                         maskim,
+                         cv2.RETR_TREE,
+                         cv2.CHAIN_APPROX_SIMPLE)
+
+    mask_polygons = []
+    for c in contours:
+        c = c.squeeze().astype('float')
+        for tf in tspec_lowres.tforms:
+            c = tf.tform(c)
+        mask_polygons.append(Polygon(c).buffer(0))
+
+    new_highres = []
+    for t in resolved_highres.tilespecs:
+        tc = t.bbox_transformed(
```

Comment:
Is there a reason to re-derive the bounding boxes here?  `t.bbox` should use the render-computed bounding boxes using points sampled by the mesh cell size along the tile width/height.

Reply:
I guess this comes down to me wanting to pass that box right into the shapely Polygon constructor which I can't do with t.bbox. It is something I've not understood about the renderapi docs which claim for t.bbox `bbox defined to fit shapely call` . Is there some alternate shapely constructor?
See below:
```
In [119]: resolved.tilespecs[0].bbox_transformed(reference_tforms=resolved.transforms)
Out[119]: 
array([[185917.99597207, 202620.56327753],
       [186070.18332292, 206458.98319249],
       [189713.86855086, 206359.78839498],
       [189708.09938615, 202399.23942917],
       [185917.99597207, 202620.56327753]])

In [120]: from shapely.geometry import Polygon

In [121]: Polygon(resolved.tilespecs[0].bbox_transformed(reference_tforms=resolved.transforms))
Out[121]: <shapely.geometry.polygon.Polygon at 0x7f705aa35190>

In [122]: resolved.tilespecs[0].bbox
Out[122]: (185917.0, 202399.0, 189713.0, 206459.0)

In [123]: Polygon(resolved.tilespecs[0].bbox)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-123-30484ae6f6bc> in <module>()
----> 1 Polygon(resolved.tilespecs[0].bbox)

/allen/programs/celltypes/workgroups/em-connectomics/danielk/conda/rm_production_mod/lib/python2.7/site-packages/shapely/geometry/polygon.pyc in __init__(self, shell, holes)
    238 
    239         if shell is not None:
--> 240             ret = geos_polygon_from_py(shell, holes)
    241             if ret is not None:
    242                 self._geom, self._ndim = ret

/allen/programs/celltypes/workgroups/em-connectomics/danielk/conda/rm_production_mod/lib/python2.7/site-packages/shapely/geometry/polygon.pyc in geos_polygon_from_py(shell, holes)
    492 
    493     if shell is not None:
--> 494         ret = geos_linearring_from_py(shell)
    495         if ret is None:
    496             return None

/allen/programs/celltypes/workgroups/em-connectomics/danielk/conda/rm_production_mod/lib/python2.7/site-packages/shapely/speedups/_speedups.pyx in shapely.speedups._speedups.geos_linearring_from_py()
    317             return None
    318 
--> 319         n = len(ob[0])
    320         if m < 3:
    321             raise ValueError(

TypeError: object of type 'float' has no len()
```